Fix MSVC possible loss of data warning in crc32_pclmulqdq.

nmoinvaz opened the pull request:
```
arch\x86\crc32_fold_pclmulqdq.c(604,43): warning C4244: 'function':
conversion from 'uint64_t' to 'size_t', possible loss of data
```

Another option we can do is to change `crc32_fold_pclmulqdq` to use `uint64_t len` instead of `size_t len`.

phprus commented:
Why the ``len`` type is ``uint64_t``? ``size_t`` can store the maximum size of a theoretically possible object of any type (including array).
Maybe it makes sense to replace ``uint64_t`` with ``size_t`` (if ``len`` is the size of the buffer in RAM)?

nmoinvaz commented:
All the crc functions use `uint64_t len`. We could change it to use `size_t` if everybody agrees.

KungFuJesus commented:
I mean the real issue here is that on 32 bit platforms, size_t is 32 bits on Windows for...goofy reasons.  I think size_t appropriately describes the size of something and I tend to like using in places that are size-like parameters.  Realistically, it's pretty unlikely that size will ever be greater than what 32 bits can contain but I suppose you could have some massive memory mapped buffer from a file you want to perform a CRC on.

KungFuJesus commented:
I think rather than doing an extra fold we make the size parameter uint64_t.  That's the expected behavior from the API, anyway.  The solution that's in the PR I don't think will work if say, the input len is 3*SIZE_MAX.  Not without a while loop in  here, that is.

nmoinvaz commented:
@KungFuJesus what do you think if we change all the crc32 functions to use sizet?

KungFuJesus commented:
> @KungFuJesus what do you think if we change all the crc32 functions to use sizet?

This creates a problem as on Windows with 32 bit platforms, size_t is uint32_t.  I'd rather it be uint64_t as that is the intent of those extended versions of the function - to handle much larger data.

mtl1979 commented:
Even on 64-bit systems, `size_t` can be 32-bit wide. That's why the warning is generated and also why we try to avoid `size_t` and `ssize_t` in our own API.

nmoinvaz commented:
Ok I will use `uint64_t` for `len`. I think it makes sense to wait for #1252 to be merged.

phprus commented:
@mtl1979
Could you please give an example of such systems?

mtl1979 commented:
> @mtl1979
> Could you please give an example of such systems?

Windows

phprus commented:
@mtl1979
Why?

A 32-bit application running on 64-bit Windows has a 32-bit address space (4 GB).
Therefore, ``size_t`` is 32 bits, since all possible objects in memory will be no larger than 4 GB.

A 64-bit application running on 64-bit Windows has a 64-bit address space and 64-bits ``size_t``.

phprus commented:
@mtl1979

Only interface functions (``functable.crc32``) have ``uint64_t len``. ``functable.crc32`` calls are made with ``size_t len``:

https://github.com/zlib-ng/zlib-ng/blob/66c6a63475f8f28b4c569d28e168ef4db94f1db2/crc32.c#L23-L35

https://github.com/zlib-ng/zlib-ng/blob/e43e0b21893b55d6b44d3f8b7105fcb655e87f5a/crc32_fold.c#L15-L27

Dead2 commented:
If we can avoid variable size parameters, I am all for it. We have also changed a lot of `int` to fixed-size data types like compared to what stock zlib has, IMO that takes some of the guesswork out of it too.

mtl1979 commented:
> @mtl1979
> Why?
> 
> A 32-bit application running on 64-bit Windows has a 32-bit address space (4 GB).
> Therefore, ``size_t`` is 32 bits, since all possible objects in memory will be no larger than 4 GB.
> 
> A 64-bit application running on 64-bit Windows has a 64-bit address space and 64-bits ``size_t``.

size_t and ssize_t are quite new additions in Visual C++, as such there has been versions where they have been declared incorrectly. Windows also use 32-bit opaque handles instead of 64-bit pointers.

mtl1979 commented:
> @mtl1979
> 
> Only interface functions (`functable.crc32`) have `uint64_t len`. `functable.crc32` calls are made with `size_t len`:

We have the externally visible `size_t` variants for compatibility reasons... All internal functions (should) use `uint64_t`. This allows us to use same code for 32-bit and 64-bit architectures. Even on 32-bit processors, 64-bit I/O can be supported.

KungFuJesus commented:
> @mtl1979
> 
> Only interface functions (`functable.crc32`) have `uint64_t len`. `functable.crc32` calls are made with `size_t len`:
> 
> https://github.com/zlib-ng/zlib-ng/blob/66c6a63475f8f28b4c569d28e168ef4db94f1db2/crc32.c#L23-L35
> 
> https://github.com/zlib-ng/zlib-ng/blob/e43e0b21893b55d6b44d3f8b7105fcb655e87f5a/crc32_fold.c#L15-L27

Upstream zlib uses long and size_t in their APIs?  Man...that's kind of a can of worms.  Specifically for the reason that long and size_t mean different things on windows if on WIN64 vs 32 bit windows.  Granted, to actually map that sort of address space like @mtl1979 was alluding to, you have to invoke mmap in a special way or use mmap64.  I guess to be compatible we have to maintain this inconsistency for amd64 vs i686 with just the windows builds.  

I do agree the fixed size types make more sense, even though size_t is an intuitive type to use for this.  If nothing else, for the fact that the 64 bit builds will want to share this function and that can become the problem exhibited above for windows.  As is, in the unlikely event that somebody manages to mmap a 64 bit address space from a file and run the CRC function on the entire width of the file, we would confusingly be limited to the first 4.2 GB, which is a bug.  The fix that's in there now to repeatedly operate on chunks of the virtual buffer isn't sufficient unless we were to loop over it to cover the potential input size the 64 bit address space allows.  Switching the type for this internal function to uint64_t fixes all of this and is more efficient.

mtl1979 commented:
> Upstream zlib uses long and size_t in their APIs? Man...that's kind of a can of worms. Specifically for the reason that long and size_t mean different things on windows if on WIN64 vs 32 bit windows. Granted, to actually map that sort of address space like @mtl1979 was alluding to, you have to invoke mmap in a special way or use mmap64. I guess to be compatible we have to maintain this inconsistency for amd64 vs i686 with just the windows builds.
> 
> I do agree the fixed size types make more sense, even though size_t is an intuitive type to use for this. If nothing else, for the fact that the 64 bit builds will want to share this function and that can become the problem exhibited above for windows. As is, in the unlikely event that somebody manages to mmap a 64 bit address space from a file and run the CRC function on the entire width of the file, we would confusingly be limited to the first 4.2 GB, which is a bug. The fix that's in there now to repeatedly operate on chunks of the virtual buffer isn't sufficiently unless we were to loop over it to cover the potential input size the 64 bit address space allows. Switching the type for this internal function to uint64_t fixes all of this and is more efficient.

Upstream zlib was written when there was still 16-bit processors, so it has a lot of weird stuff we'd rather deprecate.. Windows is getting weirder by every Visual C++ version... VC++ 2022 diverged even more from standards. For example, if something is signed on Linux, Microsoft wants it to be unsigned. Windows also use -1 (all bits set) instead of NULL for invalid pointers as NULL is reserved for user code.

KungFuJesus commented:
Hah actually now that I think about it, this doesn't even fix that, as size_t is the outer type from the API, so we'd still be limited to 4.2 GB.  Upstream really ought to address that.

mtl1979 commented on arch/x86/crc32_fold_pclmulqdq.c:
```diff
@@ -601,7 +601,12 @@ uint32_t crc32_pclmulqdq(uint32_t crc32, const unsigned char* buf, uint64_t len)
     crc32_fold ALIGNED_(16) crc_state;
     crc32_fold_reset_pclmulqdq(&crc_state);
-    crc32_fold_pclmulqdq(&crc_state, buf, len, crc32);
+    if (len > SIZE_MAX) {
```
For Windows, this should obviously be while loop...

mtl1979 commented:
> Hah actually now that I think about it, this doesn't even fix that, as size_t is the outer type from the API, so we'd still be limited to 4.2 GB. Upstream really ought to address that.

Non-compat mode should use `uint64_t` for the API too...

KungFuJesus commented:
> > Hah actually now that I think about it, this doesn't even fix that, as size_t is the outer type from the API, so we'd still be limited to 4.2 GB. Upstream really ought to address that.
> 
> Non-compat mode should use `uint64_t` for the API too...

Oh, if that's true then we are inflicting this bug on ourselves for i686 windows binaries.  In zlib-compat mode it's just a dumb limitation that feels almost like a bug.

phprus commented:
> > @mtl1979
> > Only interface functions (`functable.crc32`) have `uint64_t len`. `functable.crc32` calls are made with `size_t len`:
> 
> We have the externally visible `size_t` variants for compatibility reasons... All internal functions (should) use `uint64_t`. This allows us to use same code for 32-bit and 64-bit architectures. Even on 32-bit processors, 64-bit I/O can be supported.

64-bit I/O is not related to objects in RAM.
There are no architectures where ``size_t`` is not enough to store the size of any object in RAM.

Using ``uint64_t`` instead of ``size_t`` on 32bit architectures generates many (and slower) code. Compare assembler in  https://godbolt.org/z/cv1rE1EaP and https://godbolt.org/z/YTbE7hvKo

But all of this is not important in 64-bit architectures. In any case, the if condition ``(len > SIZE_MAX)`` from this PR or while loop condition will always be false in runtime.

KungFuJesus commented:
> 64-bit I/O is not related to objects in RAM. There are no architectures where `size_t` is not enough to store the size of any object in RAM.

That's not strictly true.  See: mmap64.  I believe you can similarly do this in Windows.

mtl1979 commented:
> 64-bit I/O is not related to objects in RAM.

That is not true... On 32-bit platform with no 64-bit I/O or LFS, it makes sense to use `uint32_t` but in all other cases it makes sense to have version with `uint64_t`.
It doesn't mean that application must use the 64-bit version if it knows the stream is or will be short enough to not overflow 32 bits.

phprus commented:
@KungFuJesus , @mtl1979
That is not true.

In the ``mmap64``, only the ``offset`` parameter has a 64-bit type.
The type of the ``length`` parameter, which specifies the length of the mapped memory region is ``size_t``. ``mmap64`` does not increase the address space available to the process. The LFS extension only makes the file size and offset type 64bit (``off_t``/``off64_t``).

See glibc source:


https://github.com/bminor/glibc/blob/f94f6d8a3572840d3ba42ab9ace3ea522c99c0c2/sysdeps/unix/sysv/linux/mmap64.c#L45-L46


and


https://github.com/bminor/glibc/blob/f94f6d8a3572840d3ba42ab9ace3ea522c99c0c2/sysdeps/unix/sysv/linux/pread64.c#L22-L23


offset - off64_t (64 bit),
len/count - size_t (32 or 64 bit)

KungFuJesus commented:
> @KungFuJesus , @mtl1979 That is not true.
> 
> In the `mmap64`, only the `offset` parameter has a 64-bit type. The type of the `length` parameter, which specifies the length of the mapped memory region is `size_t`. `mmap64` does not increase the address space available to the process. The LFS extension only makes the file size and offset type 64bit (`off_t`/`off64_t`).
> 
> See glibc source:
> 
> https://github.com/bminor/glibc/blob/f94f6d8a3572840d3ba42ab9ace3ea522c99c0c2/sysdeps/unix/sysv/linux/mmap64.c#L45-L46
> 
> and
> 
> https://github.com/bminor/glibc/blob/f94f6d8a3572840d3ba42ab9ace3ea522c99c0c2/sysdeps/unix/sysv/linux/pread64.c#L22-L23
> 
> offset - off64_t (64 bit), len/count - size_t (32 or 64 bit)

Ok, that is a fair point.  In the non-compat version we'd overflow the pointer before we ever got beyond the bounds of the 32 bit len input.  So in zlib-compat mode there's no bug and in non-compat mode we end up with a misleading argument that the user could end up doing nasty things with.  I still think making the i686 version and the amd64 version have the same API when not dealing with zlib's problem is probably a good thing?  At the very least it gives us parity with pretty much every platform that isn't windows with that regard.

mtl1979 commented:
> > 64-bit I/O is not related to objects in RAM. There are no architectures where `size_t` is not enough to store the size of any object in RAM.
> 
> That's not strictly true. See: mmap64. I believe you can similarly do this in Windows.

In Windows, people are supposed to use `VirtualAlloc` with large pages... But the original point was that we should avoid casting back and forth between 32-bit and 64-bit integer types inside our own code, only the external API wrappers should do that... So we have both 32-bit and 64-bit types in our API to support cases when LFS-aware I/O functions are used to feed the data. This allows us to count the total bytes correctly without causing wraparound. We don't actually store the whole stream in physical memory unless we explicitly enable and request large pages.

zlib-ng also has several relevant typedefs for cases when/if we absolutely need fast code on 32-bit architectures. We have already used them to overcome design flaws in API of stock zlib.

nmoinvaz committed:
Used fixed width uint8_t for crc32 and adler32 function declarations.

nmoinvaz committed:
Fix MSVC possible loss of data warning in crc32_pclmulqdq by converting len types to use uint64_t.

arch\x86\crc32_fold_pclmulqdq.c(604,43): warning C4244: 'function':
  conversion from 'uint64_t' to 'size_t', possible loss of data

nmoinvaz committed:
Use uint64_t instead of size_t for len in adler32 to be consistent with crc32.

Dead2 reviewed (approved):
LGTM

mtl1979 reviewed (approved):
LGTM.
